src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/containers/ContainerManager.java

```diff
@@ -87,170 +97,298 @@ public void setTimeoutMultiplier(int multiplier) {
-              .addPathSpecification(PathSpecification.create(containerTempDir, "/input", false))
-              .addPathSpecification(
-                  PathSpecification.create(validatorDirectory, "/validator", false))
-              .addPathSpecification(
-                  PathSpecification.create(config.getLibRoot(), "/testlib", false))
-              .addCommand("/validator/run-validator.sh")
-              .addCommand("/validator")
-              .addCommand("/testlib")
-              .addCommand("/input/input.json")
-              .setImageName(imageName)
-              .setContainerRestrictions(restrictions)
-              .setLocalUserId(config.getUid())
-              .build(),
-          containerResponse -> {
-            try {
-              return new ObjectMapper().readValue(containerResponse, ValidatorResponse.class);
-            } catch (IOException e) {
-              return new ValidatorResponse(
-                  String.format(
-                      "Failed to deserialise response from validator: %s. Response was %s.",
-                      e.getMessage(), containerResponse));
-            }
-          });
-    } catch (ContainerExecutionException | IOException e) {
-      return ContainerExecResponse.create(
-          Status.FAILED_UNKNOWN, new ValidatorResponse(e.getMessage()), e.getMessage(), -1);
     }
-  }
 
-  private static Optional<String> getMeasurements(HarnessResponse harnessResponse) {
-    ImmutableList<Measurement> m =
-        harnessResponse
-            .getTestParts()
-            .stream()
-            .map(HarnessPart::getMeasurements)
-            .flatMap(List::stream)
-            .collect(toImmutableList());
+    if (response != null && response.status() != Status.COMPLETED) {
+      callback.setStatus(Submission.STATUS_STEPS_FAILED);
+      callback.recordErrorReason(response, stepName);
+      if (response.status() != Status.ERROR) {
+        return Job.STATUS_FAILED;
+      }
+    }
+
+    callback.setStatus(Submission.STATUS_OUTPUT_RUNNING);
+
+    Execution execution = getExecution(variant, taskInfo.getOutput());
+
+    ContainerExecResponse output;
     try {
-      return Optional.of(new ObjectMapper().writeValueAsString(m));
-    } catch (JsonProcessingException e) {
-      return Optional.empty();
+      output = execOutput(
+          c.getLocation(),
+          codeDir,
+          execution,
+          variant,
+          stepResults,
+          potteryProperties);
+    } catch (ApiUnavailableException e) {
+      throw new ApiUnavailableException("Docker API unavailable when trying to execute output", e);
+    }
+
+    callback.setOutput(output.response());
+
+    if (output.status() != Status.COMPLETED) {
+      callback.setStatus(Submission.STATUS_OUTPUT_FAILED);
+      callback.recordErrorReason(response, null);
+      return Job.STATUS_FAILED;
+    }
+    return Job.STATUS_OK;
+  }
+
+  private Execution getExecution(String variant, Map<String, Execution> executionMap) {
```

Comment:
Hmmm. What do you think about @Nullable and @Nonnull? In Intellij at least, these actually give relevant warnings when you screw this up (whereas using Optional means doing an `isPresent` and then a `get`, but nothing checks the `get` has an `isPresent` in front of it, so you're relying on the `get` cluing someone in that a check is needed.)
(I'd use `ifPresent` but that can't have a break inside it so that's not good.)
What do you think?